man/detroit-email.5
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "DETROIT\-EMAIL" "5" "October 2011" "" ""
.
.SH "NAME"
\fBdetroit\-email\fR \- send out customizable email announcement
.
.SH "DESCRIPTION"
The Email plug\-in for Detroit can send out release announcements, or other messages during the standard \fBpromote\fR stop\.
.
.SH "OPTIONS"
.
.IP "\(bu" 4
\fBfile\fR \- If the body of the message is in a file use this option to set the path to that file relative to the project\'s root directory\.
.
.IP "\(bu" 4
\fBsubject\fR \- The subject of the message\. Defaults to a typical release announcement subject of "ANN: {title} {version}"\.
.
.IP "\(bu" 4
\fBmailto\fR/\fBto\fR \- List of email addresses to whom to email\.
.
.IP "\(bu" 4
\fBfrom\fR \- Email address from whom\.
.
.IP "\(bu" 4
\fBserver\fR \- Email host url\.
.
.IP "\(bu" 4
\fBport\fR \- Emails server port\. The default is usually correct\.
.
.IP "\(bu" 4
\fBaccount\fR \- Email account name (defaults to \fBfrom\fR)\.
.
.IP "\(bu" 4
\fBdomain\fR \- User domain (not sure why SMTP requires this)\.
.
.IP "\(bu" 4
\fBlogin\fR \- Login type\. Can be either \fBplain\fR or \fBlogin\fR\.
.
.IP "\(bu" 4
\fBsecure\fR \- Use TLS/SSL, true or false?
.
.IP "\(bu" 4
\fBparts\fR \- List of components of the announcement\. This will be used to construct the message as long as \fBfile\fR or \fBmessage\fR are not used\. Recognizes values are \fBreadme\fR, \fBmessage\fR, \fBdescription\fR, \fBresources\fR, \fBnotes\fR and \fBchanges\fR\.
.
.IP "\(bu" 4
\fBnoenv\fR \- Do not use environment variables for email defaults\.
.
.IP "" 0
.
.SH "DEFAULTS"
Most of the options default to environment variables (unless \fBnoenv\fR is used)\.
.
.IP "\(bu" 4
\fBEMAIL_SERVER\fR
.
.IP "\(bu" 4
\fBEMAIL_FROM\fR
.
.IP "\(bu" 4
\fBEMAIL_ACCOUNT\fR
.
.IP "\(bu" 4
\fBEMAIL_PORT\fR
.
.IP "\(bu" 4
\fBEMAIL_DOMAIN\fR
.
.IP "\(bu" 4
\fBEMAIL_LOGIN\fR
.
.IP "\(bu" 4
\fBEMAIL_SECURE\fR
.
.IP "\(bu" 4
\fBEMAIL_PASSWORD\fR
.
.IP "" 0
.
.P
Note that if EMAIL_PASSWORD is not set in the environment, then the user will be prompted for it on the command line when it is required\.
.
.SH "EXAMPLES"
Usually the use case is send out a release announcement, and in that case the default configuration suffices\. So, the typical example is as simple as:
.
.IP "" 4
.
.nf

announce:
  service: email
  mailto:
    \- ruby\-talk@ruby\-lang\.org
.
.fi
.
.IP "" 0
.
.SH "RESOURCES"
For more information:
.
.IP "\(bu" 4
API Documentation \fIhttp://rubydoc\.info/gems/detroit\-email\fR
.
.IP "\(bu" 4
Development Site \fIhttp://github\.com/detroit/detroit\-email\fR
.
.IP "" 0
.
.SH "COPYRIGHT"
Copyright (c) 2010 Thomas Sawyer, Rubyworks
.
.P
Detroit EMail is distributable in accordance with the GPLv3 license\.
.
.SH "SEE ALSO"
dnote(1)


lib/detroit-email.rb
require 'detroit/tool'

module Detroit

  # New Email instance.
  def Email(options={})
    Email.new(options)
  end

  # The Email tool is used to send out project announcements to a set of
  # email addresses.
  #
  # By default it generates a <i>release announcement</i> based on a projects
  # metadata. Release announcements can be customized via _parts_, including
  # a project's README file.
  #
  # Fully custom messages can also be sent by setting the message field,
  # and or by using a message file.
  #
  # Email account settings default to environment variables so they need
  # not be set independently for every project.
  #
  #   @server    ENV['EMAIL_SERVER']
  #   @from      ENV['EMAIL_FROM']
  #   @account   ENV['EMAIL_ACCOUNT'] || ENV['EMAIL_FROM']
  #   @password  ENV['EMAIL_PASSWORD']
  #   @port      ENV['EMAIL_PORT']
  #   @domain    ENV['EMAIL_DOMAIN']
  #   @login     ENV['EMAIL_LOGIN']
  #   @secure    ENV['EMAIL_SECURE']
  #
  class Email < Tool

    # Message file to send.
    attr_accessor :file

    # List of email addresses to whom to email.
    attr_accessor :mailto

    alias_method :to,  :mailto
    alias_method :to=, :mailto=

    # Email address from whom.
    attr_accessor :from

    # Subject line (default is "ANN: title version").
    attr_accessor :subject

    # Email server.
    attr_accessor :server

    # Emails server port (default is usually correct).
    attr_accessor :port

    # Email account name (defaults to from).
    attr_accessor :account

    # User domain (not sure why SMTP requires this?).
    attr_accessor :domain

    # Login type (plain, login).
    attr_accessor :login

    # Use TLS/SSL true or false?
    attr_reader :secure

    # Components of announcment.
    attr_accessor :parts

    # Do not use environment variables for email defaults.
    attr_accessor :noenv

    # Set if email service is using TLS/SSL security.
    def secure=(s)     
      @secure = s.to_b
    end


    #  A S S E M B L Y   S T A T I O N S

    # Attach announce method to promote assembly station.
    def station_prepare(destination)
      approve if destination == :promote
    end

    # Attach announce method to promote assembly station.
    def station_promote
      announce
    end


    #  S E R V I C E  M E T H O D S

    #
    def approve
      apply_environment
      @approved = mail_confirm?
      exit -1 unless @approved
    end

    # Send announcement message.
    def announce
      mailopts = self.mailopts

      if mailto.empty?
        report "No recipents given."
      else
        if trial?
          subject = mailopts['subject']
          mailto  = mailopts['to'].flatten.join(", ")
          report "email '#{subject}' to #{mailto}"
        else
          #emailer = Emailer.new(mailopts)
          #emailer.email
          if @approved
            email(mailopts)
          else
            exit -1
          end
        end
      end
    end

    #
    alias_method :promote, :announce

    # Message to send. Defaults to a generated release announcement.
    def message
      @message ||= (
        path = Dir[file].first if file
        if path
          project.announcement(File.new(file))
        else
          parts.map{ |part| /^file:\/\// =~ part.to_s ? $' : part }
          project.announcement(*parts)
        end
      )
    end

    # Confirm announcement
    def mail_confirm?
      if mailto
        return true if force?
        to  = [mailto].flatten.join(", ")
        ans = ask("Announce to #{to} [(v)iew (y)es (N)o]? ")
        case ans.downcase
        when 'y', 'yes'
          true
        when 'v', 'view'
          puts "From: #{from}"
          puts "To: #{to}"
          puts "Subject: #{subject}"
          puts
          puts message
          mail_confirm?
        else
          false
        end
      end
    end

    #
    def mailopts
      options = {
        'message' => self.message,
        'to'      => self.to,
        'from'    => self.from,
        'subject' => self.subject,
        'server'  => self.server,
        'port'    => self.port,
        'account' => self.account,
        'domain'  => self.domain,
        'login'   => self.login,
        'secure'  => self.secure
      }
      options.delete_if{|k,v| v.nil?}
      options
    end

    # Apply environment settings.
    def apply_environment
      return if noenv
      @server   ||= ENV['EMAIL_SERVER']
      @from     ||= ENV['EMAIL_FROM']    || ENV['EMAIL_ACCOUNT']
      @account  ||= ENV['EMAIL_ACCOUNT'] || ENV['EMAIL_FROM']
      @password ||= ENV['EMAIL_PASSWORD']
      @port     ||= ENV['EMAIL_PORT']
      @domain   ||= ENV['EMAIL_DOMAIN']
      @login    ||= ENV['EMAIL_LOGIN']
      @secure   ||= ENV['EMAIL_SECURE']
    end

    private

    def initialize_defaults
      @mailto   = ['rubytalk@ruby-lang.org']
      @subject  = "[ANN] %s v%s released" % [metadata.title, metadata.version]
      @file     = nil #'doc/ANN{,OUNCE}{,.txt,.rdoc}' TODO: default announcment file?
      @parts    = [:message, :description, :resources, :notes, :changes]

      #mailopts = Ratch::Emailer.environment_options.rekey(&:to_s)  # FIXME
      #@port    = mailopts['port']
      #@server  = mailopts['server']
      #@account = mailopts['account']  #|| metadata.email
      #@domain  = mailopts['domain']   #|| metadata.domain
      #@login   = mailopts['login']
      #@secure  = mailopts['secure']
      #@from    = mailopts['from']     #|| metadata.email
    end

    #
    def initialize_requires
      require 'facets/boolean'  # TODO: should this universal?
    end

    #
    #def announce_options(options)
    #  options  = options.rekey
    #  environ  = Emailer.environment_options
    #  defaults = project.defaults['email'].rekey
    #
    #  result = {}
    #  result.update(defaults)
    #  result.update(environ)
    #  result.update(options)
    #
    #  result[:subject] = (result[:subject] % [metadata.unixname, metadata.version])
    #
    #  result
    #end

  public

    def self.man_page
      File.dirname(__FILE__)+'/../man/detroit-email.5'
    end

  end

end
